contributing/architecture-eval-lifecycle.md

```diff
@@ -0,0 +1,326 @@
+# Architecture: Evaluations into Allocations
+
+The [Scheduling Concepts][] docs provide an overview of how the scheduling
+process works. This document is intended to go a bit deeper than those docs and
+walk thru the lifecycle of an Evaluation from job registration to running
+Allocations on the clients. The process can be broken into 4 parts:
+
+* [Job Registration](#job-registration)
+* [Scheduling](#scheduling)
+* [Deployment Watcher](#deployment-watcher)
+* [Client Allocs](#client-allocs)
+
+## Job Registration
+
+Creating or updating a Job is a _synchronous_ API operation. By the time the
+response has return to the API consumer, the Job and an Evaluation for that Job
+have been written to the state store of the server nodes.
+
+* Note: parameterized or periodic batch jobs don't create an Evaluation at
+  registration, only once dispatched.
+* Note: The scheduler is very fast! This means that once the CLI gets a
+  response it can immediately start making blocking queries to get information
+  about the next steps, but those are happening _asynchronously_.
+* Note: This workflow is different for Multi-Region Deployments (found in Nomad
+  Enterprise). That will be documented separately.
+
+The diagram below shows this initial synchronous phase. Note here that there's
+no scheduling happening yet, so no Allocation (or Deployment) has been created.
+
+```mermaid
+sequenceDiagram
+
+    participant user as User
+    participant cli as CLI
+    participant httpAPI as HTTP API
+    participant leaderRpc as Leader RPC
+    participant leaderFSM as Leader State Store
+    participant workerFSM as Worker State Store
+
+    user ->> cli: nomad job run
+    activate cli
+    cli ->> httpAPI: Create Job API
+    httpAPI ->> leaderRpc: Job.Register
+
+    activate leaderRpc
+
+    leaderRpc ->> leaderFSM: write Job
+    activate leaderFSM
+    leaderFSM ->> workerFSM: replicate Job to followers
+
+    workerFSM -->> leaderFSM: ok
+    leaderFSM -->> leaderRpc: ok
+    deactivate leaderFSM
+
+    leaderRpc ->> leaderFSM: write Evaluation
+    activate leaderFSM
+    leaderFSM ->> workerFSM: replicate Evaluation to followers
+    workerFSM -->> leaderFSM: ok
+    leaderFSM -->> leaderRpc: ok
+    deactivate leaderFSM
+
+    leaderRpc -->> httpAPI: Job.Register response
+    deactivate leaderRpc
+
+    httpAPI -->> cli: Create Job response
+    deactivate cli
+
+```
+
+## Scheduling
+
+A goroutine on the Nomad leader called the Eval Broker runs a continuous
+blocking query for Evaluations on the state store. Those Evaluations are
+enqueued in the broker.
+
+Schedulers are goroutines running on all server nodes. Typically this will be
```

Comment:
I'm not sure if it's worth mentioning this distinction here, but something that confused me early on (and I think it confused @pkazmierczak as well) is that that what we call "scheduler" in the code (like `GenericScheduler` and `SystemScheduler`) is actually just a short-lived struct that is instantiated on-demand per eval. The long-lived goroutines are the workers.

So when you're reading the code for these schedulers, they keep mutating themselves without a lock 😅

Reply:
Yeah, lemme see if I can add some language around that here.